app/rdf/query-cube-preview.ts

```diff
@@ -146,6 +151,66 @@ CONSTRUCT {
   const qsDims = qs.filter(({ p }) => p.equals(ns.sh.path));
   const dimensions: Dimension[] = [];
   const measures: Measure[] = [];
+  const observations: Observation[] = [];
+  const dimValuesByDimIri: Record<string, DimensionValue[]> = qsDims.reduce(
+    (acc, dim) => {
+      acc[dim.o.value] = [];
+      return acc;
+    },
+    {} as Record<string, DimensionValue[]>
+  );
+  // Only take quads that use dimension iris as predicates (observations values)
+  const qUniqueObservations = uniqBy(
+    qs.filter(({ p }) => qsDims.some((q) => q.o.equals(p))),
+    ({ s }) => s.value
+  );
+  qUniqueObservations.forEach(({ s }) => {
+    const sqDimValues = qsDims
+      .map((quad) => qs.find((q) => q.s.equals(s) && q.p.equals(quad.o)))
+      .filter(truthy);
+    const observation: Observation = {};
+    sqDimValues.forEach((quad) => {
+      const qDimIri = quad.p;
+      const dimIri = qDimIri.value;
+      const qDimValue = quad.o;
+      let sIri: ShortQuad | undefined;
+
+      if (!observation[dimIri]) {
+        // Retrieve the label of the observation value if it's a named node
+        if (quad.o.termType === "NamedNode") {
+          sIri = qs.find((q) => q.o.equals(quad.o));
+          const qLabel = qs.find(
+            (q) =>
+              q.s.equals(sIri?.s) &&
+              q.p.equals(qDimIri) &&
+              q.o.termType === "Literal"
+          );
+          observation[qDimIri.value] = qLabel?.o.value ?? qDimValue.value;
+        } else {
+          observation[qDimIri.value] = qDimValue.value;
+        }
+      }
+
+      const position = qs.find(
+        (q) => q.s.equals(sIri?.o) && q.p.equals(ns.schema.position)
+      )?.o.value;
+      const dimensionValue: DimensionValue = {
+        value: qDimValue.value,
+        label: `${observation[qDimIri.value]}`,
+        position: position !== undefined ? +position : undefined,
+      };
+      dimValuesByDimIri[dimIri].push(dimensionValue);
+    });
+
+    observations.push(observation);
+  });
```

Comment:
Can you index the quads by object, then subject ? Should be then more efficient than find() ?

Reply:
Done :)